### ISSUE TITLE
Ctrl C, when text is highlighted, should copy not cancel

### ISSUE THREAD

**Issue by mikemaccana** [microsoft/terminal#2285]
# Environment

WIndows 10.0.18362.0 Microsoft Windows NT 10.0.18362.0 
Windows Terminal Preview 3

# Steps to reproduce

Select text.
Press Ctrl C

# Expected behavior

Per Windows behavior, Ctrl C should copy the text.

# Actual behavior

Ctrl C cancels, potentially terminating foreground jobs. This makes sense from a terminal point of view, but Windows users are used to hitting Ctrl C to copy text. That there is some text selected is a good indication that the user wishes to copy, not to cancel.

**Comment by CobusKruger** [comment]
Yes, you're spot-on about how it should work. Unfortunately, #1093 just "fixed" this by nonsensically introducing a different keyboard shortcut for the copy function. The correct solution was mentioned several times [like here](https://github.com/microsoft/terminal/issues/968#issuecomment-495935319), but kicked to the kerb.

I can only assume it's being blocked by either ego or some internal spaghetti code.

**Comment by zadjii-msft** [comment]
@mikemaccana What do you have `copyText` bound to in your profiles.json? (you may have it bound 0, 1, or many times).

**Comment by mikemaccana** [comment]
@zadjii-msft `copyText` doesn't appear in my `profiles.json`. Want me to trash the file and let Terminal recreate it?

**Comment by zadjii-msft** [comment]
Don't do that, just add a keybinding for `copy` (that was my bad the action is `copy` not `copyText`).
```json
{ "command": "copy", "keys": ["ctrl+c"] },
{ "command": "paste", "keys": ["ctrl+v"] }
```

should do the trick

**Comment by mikemaccana** [comment]
@zadjii-msft Done - adding that keybinding stops Ctrl C from cancelling when text is selected (yaay) but breaks cancelling when text isn't selected.

**Comment by mikemaccana** [comment]
The comment mentioned earlier seems to say a lot of other people have the same idea: https://github.com/microsoft/terminal/issues/968#issuecomment-495935319

**Comment by zadjii-msft** [comment]
I'm gonna summon @carlos-zamora.

I thought we made copy only work if there was a selection, but I guess we still eat the keystroke. Maybe we shouldn't.

**Comment by Renha** [comment]
you definitely should, keyboard input shouldn't depend on mouse input - user shouldn't waste time to grab a mouse for deselection if she needs to stop some program

**Comment by mikemaccana** [comment]
Various thoughts:

 - 25 people voted for making Ctrl C copy if selected, cancel if not selected in the thread above

 - User can just bang Ctrl C and stop the program in most cases, deselection would only be needed if they explicitly selected something first. 

 - Additionally, even if they had selected something, hitting Ctrl C twice would deselect/copy and then cancel

 - We could always make it configurable.

**Comment by carlos-zamora** [comment]
As of now, we still have keystrokes eating the escape sequence. So you either have a keystroke hooked up to a command, or it emmits the escape sequence. Though it isn't ideal, the current workaround is to bind "copy" to something that won't conflict with the escape sequence you want to emit.

Though making a kind of "smart copy" would work (and I'm adding it to the coming milestone), this does show that there is a bigger issue: how can you emit escape sequences when keystrokes are bound to an action?

Accessibility was top-priority for me last month. Now that we have a somewhat working model (still plenty of work to do there), I'm hoping to spend more time shared between interactivity and accessibility. Please be patient and respectful as we make progress. We do accept PRs, as we can't get to everything between releases.

**Comment by miniksa** [comment]
> I can only assume it's being blocked by either ego or some internal spaghetti code.

@CobusKruger, please let me refresh you on the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/). We can disagree here and have debates, but please halt yourself before assuming the motivations of others and attacking individuals personally or their work.

We are happy to continue to discuss the way that this prerelease product should evolve before it meets v1.0 (and even after 1.0!) Please exercise patience and understanding when working with us and members of the community as the software develops.

**Comment by DHowett-MSFT** [comment]
Design notes:
Each key binding handler needs to be augmented with a sender (maybe) and an event args (definitely). The event args should have a `Handled()` just like the key events at the Xaml layer.
This will let the Copy/Paste/MoveToTab (etc.) handlers report whether they _did anything_ in response to the event being raised.
If a handler replies in the negatory, the event should fall through to the standard handlers.

Potential complexity:
We should be using event args to communicate _everything_ about an event. For copy, that includes "filter newlines," and for "new tab" that includes "which profile index." We may cause a bit of an explosion in event argument types.

This will tie in with @zadjii-msft's work on key binding arguments, and may be a step towards implementing it.

**Comment by zadjii-msft** [comment]
@DHowett-MSFT I'm pretty sure you just described like 80% of what #1349 has in it

**Comment by ViktorHofer** [comment]
I don't understand the downvotes in the initial post. The described expected behavior is how cmd works today in Windows. An entire booth at the #build conference celebrated when the copy cmd+c functionality was announced.

**Comment by jtleaming** [comment]
I noticed selecting text with the mouse and copying works, but if I select the text using the cursor (shift+arrow-key) and then copy, nothing seems to be copied. Maybe shift+arrow-key doesn't seem to be selecting like I thought, but the issue seems to be related.

**Comment by ryzngard** [comment]
@DHowett-MSFT  based on the design notes, should we follow #1349 to monitor this becoming available?

**Comment by Pilchie** [comment]
The lack of this paired with the poor formatting of pasting copied text from terminal into other apps has basically been a blocker to me adopting it.  The improvements in conhost of the last few years have been so great, it's been impossible to go backwards.

**Comment by CobusKruger** [comment]
@miniksa Yes, the tone of that initial response was too harsh. I've edited to tone it down but still get the point across. Hopefully the balance is better now.

I apologize for any offense caused.

**Comment by matthiasg** [comment]
As of Version 0.7.3451.0 this is not resolved. I deinstalled and reinstalled from the store and all settings were reset. Ctrl+C does not copy selected text.

**Comment by DHowett-MSFT** [comment]
@matthiasg Binding <kbd>Ctrl+C</kbd> to copy is a user preference that can be chosen by adding `{"command": "copy", "keys": ["ctrl+c"]}` to your key bindings.

**Comment by matthiasg** [comment]
I know , I did that again.

But is that really going to be the final stance on the issue? When the
terminal is released one day , then people cannot copy text from it until
they fiddle with some json file ?

I had read in some issue about this here that it was supposedly added to
new installs and only am issue with upgrades , hence my test.

**Comment by Renha** [comment]
I hope so, there is no reason to make "stop application" combination to copy text by default

**Comment by matthiasg** [comment]
@renha When text is selected (which also stops all console output normally ) ? I mean it could just copy like on a unix shell which would be even better, but windows users would expect ctrl+c after selecting text to copy text.

**Comment by DHowett-MSFT** [comment]
> could just copy

You can set`copyOnSelect` to true to get that behavior, actually. It’ll free up Ctrl+C to send Cancel if you had otherwise had it bound.

This is clearly a thorny issue.

**Comment by matthiasg** [comment]
Agreed. Obviously not ideal that Windows Shortcut for 'Copy'
 is the same as the signal shortcut to command line processes but some working defaults should be in. 

I believe once text is selected CTRL+C as copy makes sense. If someone accidentally forgets to selects text and then kills a process, I would file that as user error. It is a console after all. Can be disabled in settings but default does the right thing in both cases.

**Comment by ViktorHofer** [comment]
It's unclear to me why Ctrl+C/V isn't the default keybinding like in cmd.exe with `copyOnSelect` set to true.

**Comment by mikemaccana** [comment]
Ctrl-C cancels on most modern terminals. If you're using copyOnSelect you
don't need to copy separately.

**Comment by smflorentino** [comment]
Is there a UserVoice for Windows Terminal? I hit the same issue. It's one of those things I've just come to expect from consoles now.

**Comment by zadjii-msft** [comment]
@smflorentino _this GitHub repo_ is essentially our UserVoice. Also note that <kbd>ctrl+c</kbd> to copy when text is selected has worked for a number of releases now, _as long as you add  `{"command": "copy", "keys": ["ctrl+c"]}` to your keybindings. Are you seeing some new issue where that's not working for you?
